### ISSUE TITLE
Create a profiler package

### ISSUE THREAD

**Issue by WanWizard** [fuelphp/fuelphp#11]
We need a new profiler.

The 1.x version, which was based on PHPQuickProfiler, is no longer maintained, and it's code is complex and messy. It also lacks functionality, particularly when it comes to debugging REST API's.

The new package needs a structure not unlike Monolog. With a central logger class, a pluggable architecture for message types which unify the input to a standard internal type, a flexible message group system so messages can be grouped, and output renderers that can send the messages to a specific destination.

For both message types and renderers an interface or abstract class must be defined to type hints can be used and the presence of the correct methods is guaranteed. Renderers may be capable of outputting a message in realtime, buffer them until they are triggered, or both.

Related to #9.

**Comment by sagikazarmark** [comment]
There are quite a few other profiler packages.

**Comment by WanWizard** [comment]
Still have to find the first. Most don't really profile (like the current v1 profiler).

**Comment by sagikazarmark** [comment]
Yeah, I know, I've tried some of them myself.

**Comment by davispuh** [comment]
Have you guys found any similar loggers/profiler like PHPQuickProfiler? I'm looking for  logger/profiler library outside of FuelPHP which would be similar like PHPQuickProfiler with tabs (Console, SQL and so on) and [PSR-3](http://www.php-fig.org/psr/psr-3/) compatible. If I won't find any other I might just fork PQP from Fuel and make it as an library where you can set your own class instances for getting DB stats and such. I don't know when I'll work on that but currently I'm hoping someone have already created such library.

**Comment by WanWizard** [comment]
There is a difference between logging and profiling. Fuel uses monolog for logging, which is PSR-3 compatible, but has nothing to do with profiling or profiling data.

When it comes to profiling, you have multiple components. First, you need a container to store profiling data. The structure of this container is specific to the kind of data points you are collecting. For example, query data is completely difference from timing information or memory usage. Second, the application you are going to profile needs to provide data to the profiler. For this, some sort of unified API would be handy, probably with collector classes per type of data to be collected. And thirdly, you need an interface to display al the collected data in a form that makes sense.

PQP is I'm afraid not a really good starting point. It's codebase dates back to PHP4, the fork we use has quite a few Fuel specific modifications, is integrated into the framework, and the data points it uses are hard-coded. So to make that universally usable would be a lot of complex work.

Instead, I think it's better to start from scratch, use monolog as an architectural template, make sure you have the API and the containers for data point storage sorted, and then disect and reuse parts of PQP as a "default" rendering class.

**Comment by davispuh** [comment]
Yeah I know that. I'm specifically interested in displaying profiling and logging information on client side for developers (it would be disabled in production). Because this part would be responsible for displaying both logging and profiling like PQP does, that's why here I put both of those together. 

But yeah, it looks like that's the way to go. :)
